File: leetcode/problems/cpp/spiral-matrix.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(1)

#include "vector"
using namespace std;
class Solution {
public:
   vector<int> spiralOrder(vector<vector<int>>& matrix) {
      int n = matrix.size();
      int m = matrix[0].size();
      int left = 0, right = m - 1;
      int top = 0, bottom = n - 1;
      vector<int> ans;

      while (top <= bottom && left <= right) {
         // right is performed
         for (int i = left;i <= right;i++)
            ans.push_back(matrix[top][i]);
         top++;

         //bottom is performed
         for (int i = top;i <= bottom;i++)
            ans.push_back(matrix[i][right]);

         right--;

         // left is performed
         if (top <= bottom) {
            for (int i = right;i >= left;i--) {
               ans.push_back(matrix[bottom][i]);
            }
            bottom--;
         }

         // top is performed
         if (left <= right) {
            for (int i = bottom;i >= top;i--) {
               ans.push_back(matrix[i][left]);
            }
            left++;
         }

      }
      return ans;

   }
};

File: leetcode/problems/java/spiral-matrix.java
import java.util.List;
import java.util.ArrayList;

class Solution {
  public List<Integer> spiralOrder(int[][] matrix) {
    int left = 0, right = matrix[0].length;
    int top = 0, bottom = matrix.length;
    List<Integer> ans = new ArrayList<>();

    while (left < right && top < bottom) {
      for (int i = left; i < right; i++) {
        ans.add(matrix[top][i]);
      }
      top++;

      for (int i = top; i < bottom; i++) {
        ans.add(matrix[i][right - 1]);
      }
      right--;

      if (top < bottom) {
        for (int i = right - 1; i >= left; i--) {
          ans.add(matrix[bottom - 1][i]);
        }
        bottom--;
      }

      if (left < right) {
        for (int i = bottom - 1; i >= top; i--) {
          ans.add(matrix[i][left]);
        }
        left++;
      }
    }
    return ans;
  }
}